src/Illuminate/Support/Sort.php

```diff
@@ -0,0 +1,57 @@
+<?php
+
+namespace Illuminate\Support;
+
+class Sort
+{
+    protected $parameters = [];
+
+    /**
+     * Add a sort in ascending order using the given callback.
+     *
+     * @param  callable|string  $callback
+     * @param  int  $options
+     * @return static
+     */
+    public function asc($callback, $options = SORT_REGULAR)
+    {
+        return $this->sort($callback, $options);
+    }
+
+    /**
+     * Add a sort in descending order using the given callback.
+     *
+     * @param  callable|string  $callback
+     * @param  int  $options
+     * @return static
+     */
+    public function desc($callback, $options = SORT_REGULAR)
+    {
+        return $this->sort($callback, $options, true);
+    }
+
+    /**
+     * Add a sort using the given callback.
+     *
+     * @param  callable|string  $callback
+     * @param  int  $options
+     * @param  bool  $descending
+     * @return static
+     */
+    public function sort($callback, $options = SORT_REGULAR, $descending = false)
+    {
+        $this->parameters[] = [$callback, $options, $descending ? SORT_DESC : SORT_ASC];
+
+        return $this;
+    }
+
+    /**
+     * Get all of the sort parameters as an array.
+     *
+     * @return array
+     */
+    public function all()
+    {
+        return $this->parameters;
```

Comment:
[Tell, don't ask](https://martinfowler.com/bliki/TellDontAsk.html).

I would move the actual sorting logic into this class.

Reply:
Do you mean all of the logic (i.e. including the array_multisort call) or just the part where we map the parameters?

I considered moving the logic into here but a lot of the logic is operating on the collection itself, i.e. mapping the parameters and sorting the items. To do that I would also need to pull in the 'EnumeratesValues' trait just to get the `valueRetriever` method and that pulls in a *lot* of methods I don't think should be on this class.

I was thinking the 'sort' was more of a value object that only exists to make the API nicer, similar to how 'Fluent' is used throughout the code base. It's purpose to me seems like 'collecting sort parameters', not sorting the collection. The 'sort' doesn't have everything it needs to actually do anything; the majority of that data belongs to the collection. So it seems to me like you either need to ask the sort or the collection, and it makes more sense to ask the sort IMO since the operations are all happening on the collection's data.

Could you give an example of how you would prefer to structure it?

Reply:
I would do something like this:

```php
$items = tap(new Sort($this->items), $callback)
    ->retrieveUsing(fn ($key) => $this->valueRetriever($key))
    ->get();

return new static($items);
```

Reply:
Ok cool, thanks for the example. So it looks like you’re suggesting we make the `Sort` a general purpose class that can compound sort any array? I can definitely see how that would be generally useful but in this case it seems like the logic is too coupled to the collection.

Without a value retriever it's impossible to retrieve the columns. If you call `get` without calling `retrieveUsing` first it would throw an error. So the sort class can’t be used without providing a callback that converts a string|callable|null into a callback. In practice that means it can only be used by providing a callback that works exactly like `valueRetriever`, and that method is only used by collections.

To keep this consistent with the existing sort methods the callback should receive the value and the key, which means we have to duplicate `Collection:map` or call back into the collection for the map too.

So with a compound sort on 2 columns 4/5 operations are already provided by the collection class and we have to either use closure scope to access them, duplicate the code, or call back into the collection. That makes me think the logic should stay on the collection.

I also don't love making the `retrieveUsing` method and the `get` method available inside of the callback; it doesn't really make sense for the user to call those methods themselves but it's now part of the public API we're giving them.

I updated the PR to use the suggested API so you can see what I’m talking about. Let me know what you think.

Reply:
> Without a value retriever it's impossible to retrieve the columns. If you call `get` without calling `retrieveUsing` first it would throw an error.

I meant that by default, it should only support retrieving using a callback. Calling `retrieveUsing` should add the option to retrieve based on a key.

> To keep this consistent with the existing `sort` methods the callback should receive the value and the key, which means we have to duplicate `Collection:map` or call back into the collection for the map too.

I would've moved that functionality into our `Arr` class, and have both the `Collection` and this `sort` method call that.